kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/DataSkippingPredicate.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.skipping;
+
+import java.util.Set;
+
+import io.delta.kernel.expressions.Column;
+import io.delta.kernel.expressions.Predicate;
+
+/**
+ * Used to hold a pair of {@link Predicate} expression and {@link Set<Column>} of columns referenced
+ * by the expression.
+ */
+public class DataSkippingPredicate {
```

Comment:
why doesn't `Predicate` have a list of its own referenced columns?

something seems off that `DataSkippingPredicate` has-a `Predicate` but not is-a `Predicate`

Reply:
`Predicate` is part of our public expression framework and doesn't store referenced columns. We could rename `DataSkippingPredicate` to something else but I don't really see how it's semantically "not a `Predicate`". It's a predicate with its list of referenced columns... seems like it could be viewed as either having a predicate or being a predicate

Reply:
> but I don't really see how it's semantically "not a Predicate".

It doesn't extend Predicate, that's why. So it doesn't have a "is-a" relationship with `Predicate`. Since it's composed of a Predicate, it has a "has-a" relationship

Why doesn't it just extend `Predicate`? Every `DataSkippingPredicate` is a `Predicate`?

Reply:
Ok I did this in the latest commit LMK if that looks better to you. Alternatively we could rename `DataSkippingPredicate` to something that is more reflective of it being a pair of predicate + referenced columns.

Reply:
`DataSkippingPredicate extends Predicate` LGTM